- document Android integration (#1402)
conan-io/docs: integrations/cross_platform/android.rst

.. _android:

|android_logo| Android
____________________________

There are several ways to cross-compile packages for `Android <https://www.android.com/>`__ platform via conan.

> Using android_ndk_installer

Reviewer: make it clear that this is a Conan package. Maybe saying it is a build require directly in the title

Author: done

===========================

The easiest way so far is to use `android_ndk_installer <https://github.com/bincrafters/conan-android_ndk_installer>`_ conan package.

> Referencing mentioned package as build requirement will do the following steps:

Reviewer: Make it explicit
```suggestion
Using the ``android_ndk_installer`` package as a build requirement will do the following steps:
```

- Download the appropriate `Android NDK <https://developer.android.com/ndk>`_ archive.

- Set up required environment variables, such as ``CC``, ``CXX``, ``RANLIB`` and so on to the appropriate tools from the NDK.

> - For the CMake, will use appropriate `toolchain file <https://developer.android.com/ndk/guides/cmake#the_cmake_toolchain_file>`_ and set up necessary CMake `variables <https://developer.android.com/ndk/guides/cmake#variables>`_.

Reviewer: ```suggestion
- In case of using CMake, it will inject the appropriate `toolchain file <https://developer.android.com/ndk/guides/cmake#the_cmake_toolchain_file>`_
and set up the necessary CMake `variables <https://developer.android.com/ndk/guides/cmake#variables>`_.
```

For instance, in order to cross-compile for ``ARMv8``, the following conan profile might be used:

.. code-block:: text

  include(default)
  [settings]
  arch=armv8
  build_type=Release
  compiler=clang
  compiler.libcxx=libc++
  compiler.version=8
  os=Android
  os.api_level=21
  [build_requires]
  android_ndk_installer/r20@bincrafters/stable
  [options]
  [env]

.. note:: 

   In addition to the above, Windows users may need to specify ``CONAN_MAKE_PROGRAM``,
   for instance from the existing MinGW installation (e.g. ``C:\MinGW\bin\mingw32-make.exe``), or use make from the ``mingw_installer/1.0@conan/stable``.

Similar profile might be used to cross-compile for ``ARMv7`` (notice the ``arch`` change):

.. code-block:: text

  include(default)
  [settings]
  arch=armv7
  build_type=Release
  compiler=clang
  compiler.libcxx=libc++
  compiler.version=8
  os=Android
  os.api_level=21
  [build_requires]
  android_ndk_installer/r20@bincrafters/stable
  [options]
  [env]

> By adjusting `arch` setting, you may cross compile for ``x86`` and ``x86_64`` Android as well (e.g. if you need to run code in Sumulator).

Reviewer: ```suggestion
By adjusting ``arch`` setting, you may cross-compile for ``x86`` and ``x86_64`` Android as well (e.g. if you need to run code in Simulator).
```

.. note:: 

>   `os.api_level` is an important setting which affects compatibility - it defines ``minimum`` supported Android version.

Reviewer: ```suggestion
  ``os.api_level`` is an important setting which affects compatibility - it defines the **minimum** Android version supported.
```

.. code-block:: text

  include(default)
  target_host=aarch64-linux-android
  android_ndk=/home/conan/Library/Android/sdk/ndk/20.0.5594570
  api_level=21
  [settings]
  arch=armv8
  build_type=Release
  compiler=clang
  compiler.libcxx=libc++
  compiler.version=8
  os=Android
  os.api_level=$api_level
  [build_requires]
  [options]
  [env]
  PATH=[$android_ndk/toolchains/llvm/prebuilt/darwin-x86_64/bin]
  CHOST=$target_host
  AR=$target_host-ar
  AS=$target_host-as
  RANLIB=$target_host-ranlib
  CC=$target_host$api_level-clang
  CXX=$target_host$api_level-clang++
  LD=$target_host-ld
  STRIP=$target_host-strip

However, when building CMake projects, there are several approaches available, and it's not always clear which one to follow.

Using toolchain from Android NDK
--------------------------------

This is the official way recommended by Android developers.

For this, you will need a small CMake toolchain file:

.. code-block:: text

  set(ANDROID_PLATFORM 21)
  set(ANDROID_ABI arm64-v8a)
  include($ENV{HOME}/Library/Android/sdk/ndk/20.0.5594570/build/cmake/android.toolchain.cmake)

> This toolchain file only sets up only required CMake `variables <https://developer.android.com/ndk/guides/cmake#variables>`_, and then includes default `toolchain file <https://developer.android.com/ndk/guides/cmake#the_cmake_toolchain_file>`_ file supplied with Android NDK.

Reviewer: ```suggestion
This toolchain file only sets up the required CMake `variables <https://developer.android.com/ndk/guides/cmake#variables>`_,
and then includes the default `toolchain file <https://developer.android.com/ndk/guides/cmake#the_cmake_toolchain_file>`_ supplied with Android NDK.
```

> And then, you may use the following conan profile:

Reviewer: ```suggestion
And then, you may use the following profile:
```

.. code-block:: text

  include(default)
  [settings]
  arch=armv8
  build_type=Release
  compiler=clang
  compiler.libcxx=libc++
  compiler.version=8
  os=Android
  os.api_level=21
  [build_requires]
  [options]
  [env]
  CONAN_CMAKE_TOOLCHAIN_FILE=/home/conan/my_android_toolchain.cmake

In the profile, ``CONAN_CMAKE_TOOLCHAIN_FILE`` points to the CMake toolchain file listed above.

Using CMake build-in Android NDK support
----------------------------------------

.. warning::

>     This workflow is not supported by Android and is often broken with new NDK releases or when using older versions of CMake. Use of this workflow is strongly discouraged, and will not work with Gradle.

Reviewer: ```suggestion
    This workflow is not supported by Android and is often broken with new NDK releases or when using older versions of CMake.
    This workflow is **strongly discouraged** and will not work with Gradle.
```

> For this approach, you don't need to specify CMake toolchain file at-all. It's enough to indicate ``os`` is Android, and conan will automatically set up all required CMake `variables <https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-android>`__ for you.

Reviewer: ```suggestion
For this approach, you don't need to specify CMake toolchain file at all. It's enough to indicate ``os`` is Android
and Conan will automatically set up all required CMake
`variables <https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-android>`__ for you.
```

.. _android:

|android_logo| Android
____________________________

There are several ways to cross-compile packages for `Android <https://www.android.com/>`__ platform via conan.

Using android_ndk_installer package (build require)
=========================================================

The easiest way so far is to use `android_ndk_installer <https://github.com/bincrafters/conan-android_ndk_installer>`_ conan package.

> Using the ``android_ndk_installer`` package as a build requirement will do the following steps:

Reviewer 2: If it is not at conan-center (I haven't checked) show the command to add the remote. Otherwise comment that it is in the conan-center repository.

Author: it's in conan center, yes
